lib/IntermediateRepresentation/basicblock.h

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_INTERMEDIATEREPRESENTATION_BASICBLOCK_H_
+#define FORTRAN_INTERMEDIATEREPRESENTATION_BASICBLOCK_H_
+
+#include "mixin.h"
+#include "region.h"
+#include <iostream>
+
+namespace Fortran::IntermediateRepresentation {
+
+struct Region;
+struct Statement;
+
+struct BasicBlock final : public llvm::ilist_node<BasicBlock>,
+                          public ChildMixin<BasicBlock, Region> {
+  using StatementListType = llvm::iplist<Statement>;
+  using iterator = StatementListType::iterator;
+  using const_iterator = StatementListType::const_iterator;
+  using reverse_iterator = StatementListType::reverse_iterator;
+  using const_reverse_iterator = StatementListType::const_reverse_iterator;
+
+  BasicBlock(const BasicBlock &) = delete;
+  BasicBlock &operator=(const BasicBlock &) = delete;
+  ~BasicBlock();
+  StatementListType &getSublist(Statement *) { return Statements(); }
+  void insertBefore(Statement *stmt, Statement *before = nullptr);
+  static BasicBlock *Create(
+      Region *parentRegion, BasicBlock *insertBefore = nullptr) {
+    return new BasicBlock(parentRegion, insertBefore);
+  }
+  const Statement *getTerminator() const;
+  Statement *getTerminator() {
+    return const_cast<Statement *>(
+        const_cast<const BasicBlock *>(this)->getTerminator());
+  }
+  void SetRegion(Region *region) { parent = region; }
+  Region *GetRegion() const { return parent; }
+  void addPred(BasicBlock *bb);
+  std::vector<BasicBlock *> &Predecessors() { return predecessors_; }
```

Comment:
It would be good to standardize on either "Pred" or "Predecessor", not a mixture.

Reply:
And since this function is a simple accessor, it could be "predecessor()".

Reply:
LLVM uses `pred` but our style rules don't allow short names. It makes for nearly endless code refactoring work.

Reply:
The only reference to short names in the style guide that I can find is a preference for complete words over abbreviations.

What is the "nearly endless code refactoring work"?

Reply:
When importing code from LLVM with its own style, I have received more than a few review comments that it doesn't conform to your style guide.

Reply:
It isn't always clear that you're importing LLVM code or following an LLVM naming convention when naming something that's analogous to an LLVM object.  When you are naming something that has a direct analog in LLVM, the precedent that's in *our* style guide for reusing STL names makes sense to follow.  Since LLVM's names may not be as well known as STL names are, comments would help explain what's going on.  In the absence of comments, the names can look arbitrary.